### PR TITLE
[ML] Support for multi-parameter loss functions in derivative aggregation and finding best splits

### PR DESCRIPTION
**Pull request by tveasey**
This extends computing aggregate derivatives and solving for the highest quadratic gain split to support multi-parameter loss functions.

I experimented with various implementations. This version, which minimises the number of allocations by using memory mapped classes, proved to be much the most efficient.

This is the next step towards #982.

### CODE REVIEW

File: include/maths/CBoostedTreeLeafNodeStatistics.h [hunk 1]
```diff
@@ -37,14 +45,300 @@ class CEncodedDataFrameRowRef;
...
+
+        //! Set to the difference of \p lhs and \p rhs.
+        void subtract(const CDerivatives& rhs) {
+            m_Count -= rhs.m_Count;
+            if (m_Count > 0) {
+                m_Gradient -= rhs.m_Gradient;
+                m_Curvature -= rhs.m_Curvature;
+                // None of our loss functions have negative curvature therefore we
+                // shouldn't allow the cumulative curvature to be negative either.
+                // In this case we force it to be a v.small multiple of the magnitude
```
Comment (valeriy42) [inline include/maths/CBoostedTreeLeafNodeStatistics.h]:
```suggestion
                // In this case we force it to be a very small multiple of the magnitude
```

File: include/maths/CBoostedTreeLeafNodeStatistics.h [hunk 2]
```diff
@@ -37,14 +45,300 @@ class CEncodedDataFrameRowRef;
...
+            return m_MissingDerivatives[feature].gradient();
+        }
+
+        //! \return The aggregate curvatures for missing \p feature.
+        const TMemoryMappedDoubleMatrix& missingCurvature(std::size_t feature) const {
+            return m_MissingDerivatives[feature].curvature();
+        }
+
+        //! Add \p gradient and \p curvature to the accumulated derivatives for
+        //! the split \p split of feature \p feature.
```
Comment (valeriy42) [inline include/maths/CBoostedTreeLeafNodeStatistics.h]:
```suggestion
        //! the \p split of \p feature.
```

File: include/maths/CBoostedTreeLeafNodeStatistics.h [hunk 3]
```diff
@@ -37,14 +45,300 @@ class CEncodedDataFrameRowRef;
...
+
+    private:
+        static std::size_t number(const TDerivativesVec& derivatives) {
+            return derivatives.size();
+        }
+        static std::size_t number(const TImmutableRadixSet& splits) {
+            return splits.size() + 1;
+        }
+        template<typename SPLITS>
+        void map(const SPLITS& splits) {
```
Comment (valeriy42) [inline include/maths/CBoostedTreeLeafNodeStatistics.h]: Can you please add a comment for posterity what this function is doing? It seems to have non-trivial business logic.

File: lib/maths/CBoostedTreeLeafNodeStatistics.cc [hunk 4]
```diff
@@ -325,68 +275,134 @@ CBoostedTreeLeafNodeStatistics::computeBestSplitStatistics(const TRegularization
...
-            h += derivatives.s_Curvature;
+    // We seek to find the value at the minimum of the quadratic expansion of the
+    // regularized loss function. For a given leaf this expansion is
+    //
+    //   L(w) = 1/2 w^t H(\lambda) w + g^t w
+    //
+    // where H(\lambda) = \sum_i H_i + \lambda I, g = \sum_i g_i and w is the leaf's
+    // weight. Here, g_i and H_i denote an example's loss gradient and Hessian and i
+    // ranges over the examples in the leaf. Completing the square and noting that
+    // the minimum is given when the quadratic form is zero we have
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeLeafNodeStatistics.cc]: The last sentence is unclear: what does it mean to complete the square? And I guess, the derivative of the quadratic form should be zero.

File: lib/maths/CBoostedTreeLeafNodeStatistics.cc [hunk 5]
```diff
@@ -325,68 +275,134 @@ CBoostedTreeLeafNodeStatistics::computeBestSplitStatistics(const TRegularization
...
+    //
+    //   L(w^*) = -1/2 g^t H(\lambda) g
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeLeafNodeStatistics.cc]:
```suggestion
    //   L(w^*) = -1/2 g^t H(\lambda)^{-1} g
```
I guess, you need to invert the Hessian.

File: lib/maths/CBoostedTreeLeafNodeStatistics.cc [hunk 6]
```diff
@@ -325,68 +275,134 @@ CBoostedTreeLeafNodeStatistics::computeBestSplitStatistics(const TRegularization
...
+                splitAt = m_CandidateSplits[feature][split];
                 leftChildHasFewerRows = (2 * cl[ASSIGN_MISSING_TO_RIGHT] < c);
                 assignMissingToLeft = false;
             }
         }
 
         double penaltyForDepth{regularization.penaltyForDepth(m_Depth)};
         double penaltyForDepthPlusOne{regularization.penaltyForDepth(m_Depth + 1)};
-        double gain{0.5 * (maximumGain - CTools::pow2(g) / (h + regularization.leafWeightPenaltyMultiplier())) -
+        double gain{0.5 * (maximumGain - minimumLoss(g, h)) -
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeLeafNodeStatistics.cc]: Can you add a comment here why you store the difference?

File: lib/maths/CBoostedTreeUtils.cc [hunk 7]
```diff
@@ -6,57 +6,88 @@
...
+                       const CLoss& loss,
+                       const TMemoryMappedFloatVector& prediction,
+                       double actual,
+                       double weight) {
+    std::size_t offset{lossGradientColumn(numberInputColumns, prediction.size())};
+    auto writer = [&row, offset](std::size_t i, double value) {
+        row.writeColumn(offset + i, value);
+    };
+    // We wrap the writer in another lambda which we know takes advantage
+    // of std::function small size optimization to avoid heap allocations.
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeUtils.cc]: Good idea :+1:

File: include/maths/CBoostedTreeLeafNodeStatistics.h (hunk 3 above)
Comment (tveasey) [reply include/maths/CBoostedTreeLeafNodeStatistics.h]: Good idea, see [e1556ce](https://github.com/elastic/ml-cpp/pull/993/commits/e1556ce641f9804d8dfbb78142d0d69124ba4035).

File: include/maths/CBoostedTreeLeafNodeStatistics.h (hunk 1 above)
Comment (tveasey) [reply include/maths/CBoostedTreeLeafNodeStatistics.h]: See [e4cb871](https://github.com/elastic/ml-cpp/pull/993/commits/e4cb8710e09baaaedaf4c2669b7ead4fafcee322).

File: include/maths/CBoostedTreeLeafNodeStatistics.h (hunk 2 above)
Comment (tveasey) [reply include/maths/CBoostedTreeLeafNodeStatistics.h]: See [e4cb871](https://github.com/elastic/ml-cpp/pull/993/commits/e4cb8710e09baaaedaf4c2669b7ead4fafcee322).

File: lib/maths/CBoostedTreeLeafNodeStatistics.cc (hunk 4 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeLeafNodeStatistics.cc]: I added an extended comment and made the condition that H is positive definite clear. See [3009f42](https://github.com/elastic/ml-cpp/pull/993/commits/3009f4252f9b52adf2284b9cc06dca738ba57ab7).

File: lib/maths/CBoostedTreeLeafNodeStatistics.cc (hunk 6 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeLeafNodeStatistics.cc]: See [3009f42](https://github.com/elastic/ml-cpp/pull/993/commits/3009f4252f9b52adf2284b9cc06dca738ba57ab7).

**Comment by tveasey** [comment]
Thanks for the review @valeriy42. I've addressed all your feedback, can you take another look and let me know if you're happy?

**Pull request merged**